subsys/bluetooth/mesh/rpl_storage_settings.c

```diff
@@ -0,0 +1,279 @@
+	struct rpl_val rpl;
+	int err;
+	uint16_t src;
+
+	if (!name) {
+		BT_ERR("Insufficient number of arguments");
+		return -ENOENT;
+	}
+
+	src = strtol(name, NULL, 16);
+	entry = bt_mesh_rpl_find(src);
+
+	if (len_rd == 0) {
+		BT_DBG("val (null)");
+		if (entry) {
+			(void)memset(entry, 0, sizeof(*entry));
+		} else {
+			BT_WARN("Unable to find RPL entry for 0x%04x", src);
+		}
+
+		return 0;
+	}
+
+	if (!entry) {
+		entry = bt_mesh_rpl_alloc(src);
+		if (!entry) {
+			BT_ERR("Unable to allocate RPL entry for 0x%04x", src);
+			return -ENOMEM;
+		}
+	}
+
+	err = bt_mesh_settings_set(read_cb, cb_arg, &rpl, sizeof(rpl));
+	if (err) {
+		BT_ERR("Failed to set `net`");
+		return err;
+	}
+
+	entry->seq = rpl.seq;
+	entry->old_iv = rpl.old_iv;
+
+	BT_DBG("RPL entry for 0x%04x: Seq 0x%06x old_iv %u", entry->src,
+	       entry->seq, entry->old_iv);
+
+	return 0;
+}
+
+BT_MESH_SETTINGS_DEFINE(rpl, "RPL", rpl_set);
+
+static void store_rpl(struct bt_mesh_rpl *entry)
+{
+	struct rpl_val rpl;
+	char path[18];
+	int err;
+
+	if (!entry->src) {
+		return;
+	}
+
+	BT_DBG("src 0x%04x seq 0x%06x old_iv %u", entry->src, entry->seq,
+	       entry->old_iv);
+
+	rpl.seq = entry->seq;
+	rpl.old_iv = entry->old_iv;
+
+	snprintk(path, sizeof(path), "bt/mesh/RPL/%x", entry->src);
+
+	err = settings_save_one(path, &rpl, sizeof(rpl));
+	if (err) {
+		BT_ERR("Failed to store RPL %s value", log_strdup(path));
+	} else {
+		BT_DBG("Stored RPL %s value", log_strdup(path));
+	}
+}
+
+static void store_pending_rpl(struct bt_mesh_rpl *rpl)
+{
+	BT_DBG("");
+
+	if (atomic_test_and_clear_bit(store, rpl_idx(rpl))) {
+		store_rpl(rpl);
+	}
+}
+
+void bt_mesh_rpl_pending_store(uint16_t addr)
+{
+	int i;
+
+	if (!BT_MESH_ADDR_IS_UNICAST(addr) &&
+	     addr != BT_MESH_ADDR_ALL_NODES) {
+		return;
+	}
+
+	if (addr == BT_MESH_ADDR_ALL_NODES) {
+		bt_mesh_settings_store_cancel(BT_MESH_SETTINGS_RPL_PENDING);
+	}
+
+	for (i = 0; i < CONFIG_BT_MESH_CRPL; i++) {
+		if (addr != BT_MESH_ADDR_ALL_NODES &&
+		    addr != replay_list[i].src) {
+			continue;
+		}
+
+		if (atomic_test_bit(bt_mesh.flags, BT_MESH_VALID)) {
+			store_pending_rpl(&replay_list[i]);
+		} else {
+			clear_rpl(&replay_list[i]);
+		}
+
+		if (addr != BT_MESH_ADDR_ALL_NODES) {
+			break;
+		}
+	}
+}
+
+void bt_mesh_rpl_strg_store(struct bt_mesh_rpl *rpl, bool force)
+{
+	if (IS_ENABLED(CONFIG_BT_SETTINGS)) {
+		schedule_rpl_store(rpl, true);
+	}
+}
+
+void bt_mesh_rpl_strg_clear(struct bt_mesh_rpl *rpl)
+{
+	if (IS_ENABLED(CONFIG_BT_SETTINGS)) {
+		clear_rpl(rpl);
+	} else {
+		(void)memset(rpl, 0, sizeof(*rpl));
+	}
+}
+
+void bt_mesh_rpl_strg_clear_all(void)
+{
+	if (IS_ENABLED(CONFIG_BT_SETTINGS)) {
+		bt_mesh_settings_store_schedule(BT_MESH_SETTINGS_RPL_PENDING);
+	} else {
+		(void)memset(replay_list, 0, sizeof(replay_list));
+	}
+}
+
+int bt_mesh_rpl_get(uint16_t addr, struct bt_mesh_rpl **rpl)
+{
+	for (int i = 0; i < ARRAY_SIZE(replay_list); i++) {
+		if (!replay_list[i].src || replay_list[i].src == addr) {
```

Comment:
Will this return first free element regardless if there is an element with the addr later in the list. Could happend if clear is run.

Reply:
Can't be 100% sure, but I think this duplicates 1:1 the previous implementation for `bt_mesh_rpl_check`. The previous implementation iterated over the RPL until it found either an empty entry slot or an existing entry. As soon as either one of those criteria was fulfilled it returned. https://github.com/zephyrproject-rtos/zephyr/blob/735ddeb4a8e26c6ce07d6aa5a0b77fffc559b2a8/subsys/bluetooth/mesh/rpl.c#L124-L157

Reply:
It replicates but it seems like a bug. Some functionality requests existing RPL entry but gets an empty cell. Is this a potential attack vector for mesh devices? Especially after the reset function that cleans up some of the RPL cells.

Reply:
Right, then this is a bug in the initial implementation as well. 

Shouldn't this:

> Even more, since RPL and RPL storage are split the assumption about no empty entries in between is a kind of hidden knowledge.

be responsibility of the RPL storage?

Reply:
This is possible to solve in both:
1. Implement consideration of having empty entries in between in RPL
2. Implement some `defrag` in RPL storage to keep list without empty cells.

and probably as a separate commit or even PR.

Reply:
I'll take care of this bug.